File: fr.unice.i3s.sigma.delegates/src/fr/unice/i3s/sigma/delegates/SigmaValidationDelegateFactory.java
package fr.unice.i3s.sigma.delegates;

import static fr.unice.i3s.sigma.core.Utils.fmt;

import java.util.HashMap;
import java.util.Map;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.EClassifier;
import org.eclipse.emf.ecore.EDataType;
import org.eclipse.emf.ecore.EObject;
import org.eclipse.emf.ecore.EOperation;
import org.eclipse.emf.ecore.EValidator.ValidationDelegate;
import org.eclipse.emf.ecore.util.EObjectValidator;
import org.eclipse.emf.ecore.util.EcoreUtil;
import org.eclipse.osgi.util.NLS;

import fr.unice.i3s.sigma.core.ValidationResult;

public class SigmaValidationDelegateFactory extends
		AbstractSigmaDelegateFactory implements ValidationDelegate {

	private final Map<String, SigmaValidationDelegate> delegates = new HashMap<String, SigmaValidationDelegate>();

	@Override
	public SigmaDelegateDomain getDomain() {
		return SigmaDelegateDomain.getDefaultInstance();
	}

	protected SigmaValidationDelegate getDelegate(EClassifier eClassifier,
			String constraint) {
		String delegateId = eClassifier.getInstanceClass().getCanonicalName()
				+ "#" + constraint;
		SigmaValidationDelegate delegate = null;

		synchronized (delegates) {
			delegate = delegates.get(delegateId);
			if (delegate == null) {
				if (hasBeenDelegated(eClassifier, constraint)) {
					delegate = doCreateValidationDelegate(eClassifier,
							constraint);
				} else {
					throw new SigmaDelegateNotFoundException(
							fmt("No Sigma delegate reistered for constraint `%s` in `%s`",
									constraint, eClassifier.getName()));
				}
				delegates.put(delegateId, delegate);
			}
		}

		return delegate;
	}

	protected SigmaValidationDelegate doCreateValidationDelegate(
			EClassifier target, String constraint) {
		return new SigmaValidationDelegate(target, getDomain(), constraint);
	}

	protected boolean hasBeenDelegated(EClassifier target, String constraint) {
		return EcoreUtil
				.getAnnotation(target, getDomain().getURI(), constraint) != null;
	}

	// following methods are used from the plain EObjectValidator

	/**
	 * This method is consequently used by the {@link SigmaEObjectValidator} so
	 * it has access to the full results.
	 */
	private ValidationResult doValidate(EClassifier eClassifier, Object object,
			Map<Object, Object> context, String constraint, String expression) {

		try {
			SigmaValidationDelegate delegate = getDelegate(eClassifier,
					constraint);
			return delegate.validate(eClassifier, object, constraint,
					expression);
		} catch (SigmaDelegateNotFoundException e) {
			// TODO log
			e.printStackTrace();

			String message = NLS.bind(
					Messages.Sigma_NoSigmaRegisteredValidationDelegate,
					new Object[] { constraint,
							getLabel(eClassifier, object, context) });

			return ValidationResult.error(message);
		}
	}

	private static String getLabel(EClassifier eClassifier, Object value,
			Map<Object, Object> context) {
		if (value instanceof EObject) {
			return EObjectValidator.getObjectLabel(eClassifier, context);
		} else {
			return EObjectValidator.getValueLabel((EDataType) eClassifier,
					value, context);
		}
	}

	@Override
	public boolean validate(EClass eClass, EObject eObject,
			Map<Object, Object> context, EOperation invariant, String expression) {
		return doValidate(eClass, eObject, context, invariant.getName(),
				expression).isValidOrCanceled();
	}

	@Override
	public boolean validate(EClass eClass, EObject eObject,
			Map<Object, Object> context, String constraint, String expression) {
		return doValidate(eClass, eObject, context, constraint, expression)
				.isValidOrCanceled();
	}

	@Override
	public boolean validate(EDataType eDataType, Object value,
			Map<Object, Object> context, String constraint, String expression) {
		return doValidate(eDataType, value, context, constraint, expression)
				.isValidOrCanceled();
	}

}

File: fr.unice.i3s.sigma/src/fr/unice/i3s/sigma/delegates/SigmaValidationDelegateFactory.scala
package fr.unice.i3s.sigma.delegates

import scala.collection.JavaConversions._
import org.eclipse.emf.ecore.EClass
import org.eclipse.emf.ecore.EValidator.ValidationDelegate
import org.eclipse.emf.ecore.EDataType
import org.eclipse.emf.ecore.EObject
import org.eclipse.emf.ecore.EOperation
import java.util.{ Map ⇒ JMap }
import org.eclipse.emf.ecore.scala.EcorePackageScalaSupport
import org.eclipse.emf.ecore.util.EcoreUtil
import java.lang.reflect.Method
import fr.unice.i3s.sigma.validation.Error
import org.eclipse.emf.ecore.util.EObjectValidator
import fr.unice.i3s.sigma.validation.ValidationResult

final class SigmaValidationDelegateFactory extends AbstractSigmaDelegateFactory[(EClass, String)] with ValidationDelegate with EcorePackageScalaSupport {

  protected val delegates = collection.mutable.Map[String, SigmaValidationDelegate]()

  protected def getDelegate(clazz: EClass, constraint: String) = {
    val id = clazz.instanceClass.getCanonicalName + "#" + constraint
    delegates.synchronized {
      delegates.get(id) match {
        case Some(delegate) ⇒ delegate
        case None ⇒ {
          Option(createValidationDelegate(clazz, constraint)) match {
            case Some(delegate) ⇒
              delegates += (id -> delegate)
              delegate
            case None ⇒
              throw new SigmaDelegateNotFoundException(
                s"No Sigma delegate registered for constraint `$constraint` in `${clazz.name}`")
          }
        }
      }
    }
  }

  protected def createValidationDelegate(clazz: EClass, constraint: String) = {
    if (isDelegated(clazz, constraint)) doCreateDelegate(clazz, constraint)
    else null
  }

  protected def doCreateDelegate(clazz: EClass, constraint: String) = {
    lookupDelegate((clazz, constraint), clazz, Option(EcoreUtil.getAnnotation(clazz, domain.getURI, constraint))) match {
      case Right(delegateMethod) ⇒
        new SigmaValidationDelegate(clazz, constraint, delegateMethod, domain)
      case Left(msg) ⇒
        throw new SigmaDelegateNotFoundException(s"Unable to find invocation delegate for ${clazz.getName}#$constraint: $msg")
    }
  }

  // TODO: implement
  protected def isDelegated(clazz: EClass, constraint: String) = true

  protected def doValidate(clazz: EClass, target: EObject, context: JMap[Object, Object], constraint: String): ValidationResult =
    try {
      val delegate = getDelegate(clazz, constraint)
      delegate.validate(target)
    } catch {
      case e: SigmaDelegateNotFoundException ⇒
        Error(s"No Sigma delegate has been found for constraint `$constraint` for `$target`")
      case e: Throwable ⇒
        Error(s"Error in executing Sigma validation delegate for constraint `$constraint` for `$target`: ${e.getMessage}")
    }

  override def delegateMethodName(element: (EClass, String)): String = "validate" + element._2

  override def checkMethodSignature(element: (EClass, String), method: Method): Boolean = {
    // must have a compatible return type
    if (!method.getReturnType.isAssignableFrom(classOf[Boolean])) false

    val mparams = method.getGenericParameterTypes

    // must have the first argument of type of the operation container
    if (!(mparams.length == 1 &&
      domain.compatibleType(element._1, mparams(0)))) false

    true
  }

  override def methodSignature(element: (EClass, String)): String = {
    val selfType = domain.classifierTypeName(element._1)
    val name = delegateMethodName(element)
    val typeParams = domain.classifierTypeParameters(element._1)

    s"def $name$typeParams(self: $selfType): Boolean"
  }

  override def validate(eClass: EClass, eObject: EObject,
    context: JMap[Object, Object], invariant: EOperation, expression: String) =
    doValidate(eClass, eObject, context, invariant.getName()).isValidOrCanceled;

  override def validate(eClass: EClass, eObject: EObject,
    context: JMap[Object, Object], constraint: String, expression: String) =
    doValidate(eClass, eObject, context, constraint)
      .isValidOrCanceled

  override def validate(eDataType: EDataType, value: Object,
    context: JMap[Object, Object], constraint: String, expression: String) =
    // TODO: better exception
    throw new RuntimeException;

}